File: leetcode/leetcode_cpp/find-pivot-index.cpp
/*
https://leetcode.com/problems/find-pivot-index

1. Sum list and compare
time: o(n)
space: o(n)

2. Accumulate and distract in iteration
time: o(n)
space: o(1)
*/
class Solution {
public:
    int pivotIndex_1(vector<int>& nums) {
        if (!nums.size()) return -1;
        vector<int> nums_sum(nums.size(), 0);
        int s = 0;
        for (int i=nums.size()-1; i>=0; --i) {
            nums_sum[i] = s;
            s += nums[i];
        }
        // 0 1 8 11 17 22 28
        // 28 27 20 17 11 6 0
        
        s = 0;
        for (int i=0; i<nums.size(); ++i) {
            if (s == nums_sum[i])
                return i;
            s += nums[i];
        }
        return -1;
    }
    int pivotIndex_2(vector<int>& nums) {
        if (!nums.size()) return -1;
        int s = 0;
        for (int& num : nums) s += num;

        int s2 = 0;
        for (int i=0; i<nums.size(); ++i) {
            s -= nums[i];
            if (s2 == s) return i;
            s2 += nums[i];
        }

        return -1;
    }
    int pivotIndex(vector<int>& nums) {
        return pivotIndex_2(nums);
    }
};

File: leetcode/leetcode_py/find-pivot-index.py
'''
https://leetcode.com/problems/find-pivot-index

1. Sum list and compare
time: o(n)
space: o(n)

2. Accumulate and distract in iteration
time: o(n)
space: o(1)
'''
class Solution:
    def pivotIndex_1(self, nums: List[int]) -> int:
        v1 = [0] * len(nums)
        s = 0
        for i in range(len(nums)-1,-1,-1):
            v1[i] = s
            s += nums[i]

        s = 0
        for i in range(len(nums)):
            if s == v1[i]:
                return i
            s += nums[i]
        return -1

    def pivotIndex_2(self, nums: List[int]) -> int:
        s = 0
        for num in nums: s += num

        s1 = 0
        for i in range(len(nums)):
            s -= nums[i]
            if s == s1:
                return i
            s1 += nums[i]
        return -1

    def pivotIndex(self, nums: List[int]) -> int:
        return self.pivotIndex_2(nums)
        